File: algorithms/628. Maximum Product of Three Numbers/solution.cc
class Solution {
public:
    int maximumProduct(vector<int>& nums) {
        int n1 = 1001, n2 = 1001, m1 = -1001, m2 = -1001, m3 = -1001;
        for ( int n : nums ) {
            if ( n >= m1 ) {
                m3 = m2;
                m2 = m1;
                m1 = n;
            } else if ( n >= m2 ) {
                m3 = m2;
                m2 = n;
            } else if ( n > m3 ) {
                m3 = n;
            }

            if ( n <= n1 ) {
                n2 = n1;
                n1 = n;
            } else if ( n < n2 ) {
                n2 = n;
            }
        }
        return max(n1 * n2 * m1, m1 * m2 * m3);
    }
};

// 83 / 83 test cases passed.
// Status: Accepted
// Runtime: 49 ms

File: algorithms/628. Maximum Product of Three Numbers/solution.py
class Solution(object):
    def maximumProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n1 = n2 = float('inf')
        m1 = m2 = m3 = float('-inf')
        for n in nums:
            if n >= m1:
                m1, m2, m3 = n, m1, m2
            elif n >= m2:
                m2, m3 = n, m2
            elif n > m3:
                m3 = n
            if n <= n1:
                n1, n2 = n, n1
            elif n < n2:
                n2 = n
        return max(n1 * n2 * m1, m1 * m2 * m3)

# 83 / 83 test cases passed.
# Status: Accepted
# Runtime: 83 ms
